src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -32,71 +34,88 @@ export default class GooglePayPaymentStrategy implements PaymentStrategy {
-        if (this._walletButton) {
-            this._walletButton.removeEventListener('click', this._handleWalletButtonClick);
+        if (this._walletButton && this._buttonClickEventHandler) {
+            this._walletButton.removeEventListener('click', this._buttonClickEventHandler);
         }
 
+        this._buttonClickEventHandler = undefined;
         this._walletButton = undefined;
 
         return this._googlePayPaymentProcessor.deinitialize()
             .then(() => this._store.getState());
     }
 
-    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+    async execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
         if (!this._googlePayOptions) {
             throw new InvalidArgumentError('Unable to initialize payment because "options.googlepay" argument is not provided.');
         }
 
-        const {
-            onError = () => {},
-            onPaymentSelect = () => {},
-        } = this._googlePayOptions;
-
-        return Promise.resolve(this._getPayment())
-            .then(payment => {
-                if (!payment.paymentData.nonce || !payment.paymentData.cardInformation) {
-                    // TODO: Find a way to share the code with _handleWalletButtonClick method
-                    return this._googlePayPaymentProcessor.displayWallet()
-                        .then(paymentData => this._paymentInstrumentSelected(paymentData))
-                        .then(() => onPaymentSelect())
-                        .then(() => this._getPayment())
-                        .catch(error => {
-                            if (error.statusCode !== 'CANCELED') {
-                                onError(error);
-                            }
-                        });
-                }
-
-                return payment;
-            })
-            .then(() =>
-                this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit: payload.useStoreCredit }, options))
-                    .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment(this._getPayment())))
-                    .catch(error => this._googlePayAdyenV2PaymentProcessor?.processAdditionalAction(error) || Promise.reject(error))
-            );
+        if (!payload.payment) {
+            throw new PaymentArgumentInvalidError(['payment']);
+        }
+
+        const { methodId } = payload.payment;
+
+        let payment = await this._getPayment(methodId);
+
+        if (!payment.paymentData.nonce || !payment.paymentData.cardInformation) {
+            const {
+                onError = () => {},
+                onPaymentSelect = () => {},
+            } = this._googlePayOptions;
+            await this._displayWallet(methodId, onPaymentSelect, onError);
+            payment = await this._getPayment(methodId);
+        }
+
+        if (!payment.paymentData.nonce) {
```

Comment:
this `if` condition makes sense to happen inside the previous `if` block. This is to be explicit around how the logic will be executed

```js
        if (!payment.paymentData.nonce || !payment.paymentData.cardInformation) {
            const {
                onError = () => {},
                onPaymentSelect = () => {},
            } = this._googlePayOptions;
            await this._displayWallet(methodId, onPaymentSelect, onError);
            payment = await this._getPayment(methodId);

            if (!payment.paymentData.nonce) {
              throw new MissingDataError(MissingDataErrorType.MissingPayment);
            }
        }
```